### ISSUE TITLE
Add support for CTEs

### ISSUE THREAD

**Issue by beikov** [Blazebit/blaze-persistence#157]
Apparently no persistence provider supports CTEs yet but I have an idea how I could support them.
Basically you have a criteria builder just as always, but additionally you can add some CTEs. Rendering the SQL and doing the parameter bindings shouldn't be that hard but will be provider specific. I don't know yet how I will implement the referencing of a CTE but I guess there will be some overloads of the ```from```-method.

**Comment by beikov** [comment]
Started working on this and the only possible way to make that work smoothly is to integrate with the persistence provider on a deeper level. To make CTEs available for a persistence provider, it needs to support mapping read only entities. The schema generation may not generate DDL code for it and there should be a possibility to define the relation name. It must be queryable and therefore also be included in the metamodel.

**Comment by beikov** [comment]
Note that CTEs kind of require multiple from clause elements to work

**Comment by beikov** [comment]
Fixed in 1.1.0 by https://github.com/Blazebit/blaze-persistence/commit/e6057e3e3484b5a7a86fd71be7bd93b311c21d8f

**Comment by beikov** [comment]
Here is a nice overview of the supported databases: http://stackoverflow.com/questions/324935/mysql-with-clause

**Comment by TuomasKiviaho** [comment]
> Here is a nice overview of the supported databases: http://stackoverflow.com/questions/324935/mysql-with-clause

Yesterday I got blaze-persistence up and running on [OSGi](https://osgi.org/specification/osgi.enterprise/7.0.0/service.loader.html) and hooked it successfully with Hibernate. Of course as usual I studied the docs afterwards and noticed that MySQL is not well supported on the CTE side.

I could neither find any information from github nor from stackoverflow whether or not the mentioned WITH clause could be supported when using a modern MySql as the update of the overview states.
